### DISCUSSION TITLE
Questions About Implementing ara::com with SOME/IP

### DISCUSSION THREAD

**Discussion by ShehabAldeenMo** [langroodi/Adaptive-AUTOSAR#40] [Q&A]
Dear Engineer Langroodi,

For my graduation project, I’ll be focusing on implementing Adaptive AUTOSAR clusters, starting with ara::com. However, I need some clarification on the implementation details.

Referring to the image below and my current understanding:
![Screenshot from 2024-12-12 01-13-38](https://github.com/user-attachments/assets/a2549be9-8a55-4eb4-bede-9aa8e19a4c73)
I’m unsure whether I need to implement SOME/IP under the ara::com cluster or if it’s possible to leverage existing solutions like [vsomeip](https://github.com/COVESA/vsomeip). While I understand that vsomeip isn’t exactly SOME/IP, I’d appreciate your insights on the following questions:

Questions:-
--------------
1. Implementation or Reuse: Do I need to implement SOME/IP from scratch, or can I reuse an existing solution? If reusing is an option, where can I find the appropriate resources?

2. Starting Point in ara::com: From your perspective, which part of ara::com would be the best starting point for someone diving into this project?

---

Thank you in advance for your guidance!

**Comment by langroodi** [comment]
Dear Shehab,

> Implementation or Reuse: Do I need to implement SOME/IP from scratch, or can I reuse an existing solution? If reusing is an option, where can I find the appropriate resources?

It is not clear to me how you come to the conclusion that **vsomeip** is not exactly SOME/IP. As far as they (COVESA) claim, it is the implementation of SOME/IP. So, you can use it without any concern. To start with, please check [their use guide](https://github.com/COVESA/vsomeip/blob/master/documentation/vsomeipUserGuide.md) and also [their examples](https://github.com/COVESA/vsomeip/tree/master/examples) (specially the **hello_word** project).

> Starting Point in ara::com: From your perspective, which part of ara::com would be the best starting point for someone diving into this project?

To be clear, ara::com it is not only SOME/IP, but it also can be DDS (as also mentioned in the image you've shared). On top of these communication protocols, we have the ara::com API. So, you can start with [this document](https://www.autosar.org/fileadmin/standards/R24-11/AP/AUTOSAR_AP_EXP_ARAComAPI.pdf) to understand the concept of **Proxy/Skeleton Architecture** and then learn more about the API itself.
At some point, you need to make the decision how many of these APIs (for instance, shall you have a service discovery or not) and how many of these communication protocols (only SOME/IP or SOME/IP + DDS) you want to cover in your project. If you want to cover the whole thing, maybe then it is more than a graduation project and the added value is also questionable.
